component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
+                            this.clientStatus = ClientStatus.RETRYING;
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, false);
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else if (e instanceof AmqpException && ((AmqpException) e).getErrorCondition()
+                                == AmqpErrorCondition.RESOURCE_LIMIT_EXCEEDED) {
+                            //If resource limit is exceeded we will retry after a constant delay
+                            log.error("Resource limit exceeded when publishing EventDataBatch. Operation will be "
+                                              + "retried after constant delay");
+                            try {
+                                Thread.sleep(1000 * 60);
+                            } catch (InterruptedException interruptedException) {
+                                Thread.currentThread().interrupt();
+                            }
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else {
+                            //For any other exception
+                            log.error("Unknown error occurred while publishing EventDataBatch. Producer client will "
+                                              + "be re-initialized. Events may be lost in the process.");
+                            this.clientStatus = ClientStatus.RETRYING;
+                            readWriteLock.writeLock().unlock();
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, true);
+                        }
+                    } finally {
+                        try {
+                            readWriteLock.writeLock().unlock();
+                        } catch (IllegalMonitorStateException e) {
+                            //ignore
+                        }
                     }
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                } finally {
-                    sendSemaphore.release();
+                }
+                if (isAdded) {
+                    log.debug("Adding event: " + event.replaceAll("[\r\n]", ""));
+                } else {
+                    log.debug("Failed to add event: " + event.replaceAll("[\r\n]", ""));
+                }
+            } finally {
+                try {
+                    readWriteLock.readLock().unlock();
+                } catch (IllegalMonitorStateException e) {
+                    //ignore
                 }
             }
+        } else {
+            try {
+                threadBarrier.lock();
+                log.debug(Thread.currentThread().getName().replaceAll("[\r\n]", "") + " will be parked as EventHub "
+                                  + "Client is inactive.");
+                waitCondition.await();
```

Comment:
Shouldn't we put a timeout here? Otherwise indefinitely waiting on the waitCondition would exhaust worker threads too right?

Reply:
We can release. But then thread would go back to task queue take a task and again wain on same condition